File: linked_list/merge/cpp/merge_k_sorted_lists.cpp
// Merge K Sorted Lists — 合并 K 个有序链表 (Hard)
// Approach: 迭代分治（两两合并，间隔倍增）/ Iterative divide-and-conquer (pairwise merge with doubling interval)
// Time: O(N log K)  (N 为所有节点总数, K 为链表个数)
// Space: O(1) 额外空间（不含返回链表）/ O(1) extra (excluding result list)

#include <bits/stdc++.h>
using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};
// mergeTwo: 合并两个有序链表，返回新链表头 / Merge two sorted lists and return the merged head
ListNode *mergeTwo(ListNode *a, ListNode *b)
{
    ListNode dummy(0);
    ListNode *tail = &dummy;
    while (a && b)
    {
        if (a->val <= b->val)
        {
            tail->next = a;
            a = a->next;
        }
        else
        {
            tail->next = b;
            b = b->next;
        }
    }
    tail->next = a ? a : b;
    return dummy.next;
}

// 迭代分治合并 K 个链表 / Divide-and-conquer (iterative) merge K lists
ListNode *mergeK(vector<ListNode *> &lists)
{
    if (lists.empty())
        return nullptr;
    int n = (int)lists.size();
    int interval = 1;
    while (interval < n)
    {
        for (int i = 0; i + interval < n; i += interval * 2)
        {
            lists[i] = mergeTwo(lists[i], lists[i + interval]);
        }
        interval *= 2;
    }
    return lists[0];
}

// helpers for demo
ListNode *build(const vector<int> &v)
{
    ListNode dummy(0);
    ListNode *t = &dummy;
    for (int x : v)
    {
        t->next = new ListNode(x);
        t = t->next;
    }
    return dummy.next;
}

void print(ListNode *h)
{
    while (h)
    {
        cout << h->val;
        if (h->next)
            cout << " -> ";
        h = h->next;
    }
    cout << "\n";
}

int main()
{
    vector<ListNode *> lists;
    lists.push_back(build({1, 4, 5}));
    lists.push_back(build({1, 3, 4}));
    lists.push_back(build({2, 6}));

    ListNode *res = mergeK(lists);
    print(res);
    return 0;
}

File: linked_list/merge/lua/merge_k_sorted_lists.lua
-- Merge K Sorted Lists — 合并 K 个有序链表 (Hard)
-- 方法: 迭代分治（两两合并，间隔倍增）/ Iterative divide-and-conquer (pairwise merge with doubling interval)
-- 时间复杂度: O(N log K)，空间复杂度: O(1) 额外空间（不计返回链表）

local ListNode = {}
ListNode.__index = ListNode
function ListNode.new(val)
    return setmetatable({val = val, next = nil}, ListNode)
end

-- mergeTwo: 合并两个有序链表 / merge two sorted lists
local function mergeTwo(a, b)
    local dummy = ListNode.new(0)
    local tail = dummy
    while a ~= nil and b ~= nil do
        if a.val <= b.val then
            tail.next = a
            a = a.next
        else
            tail.next = b
            b = b.next
        end
        tail = tail.next
    end
    tail.next = a or b
    return dummy.next
end

-- mergeK: 通过倍增间隔两两合并 K 个链表 / merge K lists by doubling interval
local function mergeK(lists)
    local n = #lists
    if n == 0 then
        return nil
    end
    local interval = 1
    while interval < n do
        local i = 1
        while i + interval <= n do
            lists[i] = mergeTwo(lists[i], lists[i + interval])
            i = i + interval * 2
        end
        interval = interval * 2
    end
    return lists[1]
end

-- build: 由数组构造链表 / build list from array
local function build(arr)
    local dummy = ListNode.new(0)
    local t = dummy
    for _, v in ipairs(arr) do
        t.next = ListNode.new(v)
        t = t.next
    end
    return dummy.next
end

-- printList: 打印链表 / print list
local function printList(h)
    local first = true
    while h do
        if not first then
            io.write(" -> ")
        end
        io.write(tostring(h.val))
        first = false
        h = h.next
    end
    io.write("\n")
end

-- Demo: 构造三个有序链表并合并 / Build and merge 3 sorted lists
local lists = {
    build({1, 4, 5}),
    build({1, 3, 4}),
    build({2, 6})
}
local res = mergeK(lists)
printList(res)
